src/midgard/polyline2.cc

```diff
@@ -65,59 +65,79 @@ std::tuple<coord_t, float, int> Polyline2<coord_t>::ClosestPoint(const coord_t&
   // method to generalize the polyline. Square the error tolerance to avoid
   // sqrts.
-  Generalize(pts_, t);
+  Generalize(pts_, t, indices);
   return pts_.size();
 }
 
 // Get a generalized polyline from this polyline. This polyline remains
 // unchanged.
-template <class coord_t> Polyline2<coord_t> Polyline2<coord_t>::GeneralizedPolyline(const float t) {
+template <class coord_t>
+Polyline2<coord_t>
+Polyline2<coord_t>::GeneralizedPolyline(const float t, const std::unordered_set<size_t>& indices) {
   // Recursively call Douglass-Peucker method to generalize the polyline.
   // Square the error tolerance to avoid sqrts.
   Polyline2 generalized(pts_);
-  generalized.Generalize(t);
+  generalized.Generalize(t, indices);
   return generalized;
 }
 
 template <class coord_t>
 template <class container_t>
-void Polyline2<coord_t>::Generalize(container_t& polyline, float epsilon) {
-
+void Polyline2<coord_t>::Generalize(container_t& polyline,
+                                    float epsilon,
+                                    const std::unordered_set<size_t>& indices) {
   // the recursive bit
   epsilon *= epsilon;
-  std::function<void(typename container_t::iterator, typename container_t::iterator)> peucker;
-  peucker = [&peucker, &polyline, epsilon](typename container_t::iterator start,
-                                           typename container_t::iterator end) {
+  std::function<void(typename container_t::iterator, size_t, typename container_t::iterator, size_t)>
+      peucker;
+  peucker = [&peucker, &polyline, epsilon, &indices](typename container_t::iterator start, size_t s,
+                                                     typename container_t::iterator end, size_t e) {
     // find the point furthest from the line
     float dmax = 0.f;
     typename container_t::iterator itr;
     LineSegment2<coord_t> l{*start, *end};
+    size_t j = e - 1, k;
     coord_t tmp;
-    for (auto i = std::next(start); i != end; ++i) {
+    for (auto i = std::prev(end); i != start; --i, --j) {
+      // special points we dont want to generalize no matter what take precidence
+      if (indices.find(j) != indices.end()) {
+        itr = i;
+        dmax = epsilon;
+        k = j;
+        break;
+      }
+
+      // if this is the highest frequency detail so far
       auto d = l.DistanceSquared(*i, tmp);
       if (d > dmax) {
         itr = i;
         dmax = d;
+        k = j;
       }
     }
 
     // there are some high frequency details between start and end
     // so we need to look for flatter sections between them
     if (dmax >= epsilon) {
-      peucker(start, itr);
-      peucker(itr, end);
-    } // nothing sticks out between start and end so simplify it away
-    else {
+      // we recurse from right to left for two reasons:
+      // 1. we want to preserve iterator validity in the vector version
+      // 2. its the only way to preserve the indices in the keep set
+      if (e - k > 1)
```

Comment:
here and above in the loop we make sure to go from right to left when generalizing so that we dont invalidate iterators in vector containers. this wasnt a problem in isochrones because its using a list but going in this direction also has the benefit of being able to mark specific indices of a polyline to not be generalized without jumping through lots of hoops.